design/userdefined-rootdefinehints.md

```diff
@@ -0,0 +1,101 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# userdefined-rootdevicehints
+
+## Status
+
+provisional
+
+## Table of Contents
+
+<!--ts-->
+   * [userdefined-rootdevicehints](#userdefined-rootdevicehints)
+      * [Status](#status)
+      * [Table of Contents](#table-of-contents)
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+         * [Goals](#goals)
+         * [Non-Goals](#non-goals)
+      * [Proposal](#proposal)
+         * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Design Details](#design-details)
+         * [Work Items](#work-items)
+         * [Dependencies](#dependencies)
+         * [Test Plan](#test-plan)
+         * [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+         * [Version Skew Strategy](#version-skew-strategy)
+      * [References](#references)
+
+<!-- Added by: dhellmann, at: Tue Jun 25 14:54:44 EDT 2019 -->
+
+<!--te-->
+
+## Summary
+
+This document explains how rootdevicehints could be given as user defined parameters as part BaremetalHostSpec. The idea is that user would have a possibilty to defines selective constrains for root device selection.
+
+## Motivation
+
+We need to be able select root device, when there is multiple devices available so that device is not chosen randomly
+
+### Goals
+
+1. Decide root device hints are needed be supported.
+2. Agree on the way haw information is propagated to ironic through BMO
+
+
+### Non-Goals
+
+The complete set of root device hints as input is not needed only those seen important.
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+- BaremetalHostSpec need to be updated to support list of following root device hints:
+* vendor
+* size
+* wwn
+
+The above parameters are optional in BaremetalHostSpec. If user is using legacy Hardware profiles, the root device hints hardcoded to code can be overwritten or appended with proposed root device hints
+
+### Risks and Mitigations
+
+-
+
+## Design Details
+
+### Work Items
+
+- Update BaremetalHostSpec with new RootDeviceHints type struct to have above parameters as optional.
```

Comment:
If we put the hints in a struct, then we need to explain the precedence order that will be used in case the user gives us multiple values. For example, if they give HCTL and WWN, they may not refer to the same device. Do we use both of them and expect them to match? Or do we say that only 1 of the values is used and provide the order in which they are checked?

Is there a way to express that decision in the API so that the user won't be surprised in the outcome?

Reply:
Our use cases are such that sometimes we would like to use deterministic selector like WWN and sometime we would like to get any device that comply for ex. vendor and size. In other words our need are such that user needs to know what it want's and not mixing deterministic ids with group selectors. Should we made a logic in BMO which would ignore group selectors in case there is deterministic selector like WWN?

Reply:
What is setting the values on the BareMetalHost in your use case? Is it the cluster API provider? Or something else?

Reply:
We use wrapper scripts, maybe later clusterctl or metalctl(idea)

Reply:
OK, as I mention above, I think we want the BareMetalHost API to be deterministic and any smart logic to be in the thing that invokes it.